Add proposal for interest bearing tokens (#15927)
solana-labs/solana: docs/src/proposals/interest-bearing-tokens.md

---
title: Interest-Bearing Tokens
---

This document describes a solution for interest-bearing tokens as an on-chain

> program. Interest-bearing tokens continuously accrue interest as time passes.

Reviewer: i-tokens are also extensible for tokens with lockup and more customized tokens?

Author: You would probably create another token program for those special cases.  This i-token proposal is really two things:

* the i-token program, which is just for interest-bearing tokens
* updating the overall programming model to suit N token programs

So to answer your question, I wouldn't expand the i-tokens for lockup / more customization, but once the programming model is updated, it'll be possible to create another token program for things like lockup.

### How do these work?

Let's explain through an example: assume there's an i-token mint with a total
supply of 1,000 tokens
and a 10% annual interest rate. This means that one year later, the total supply
will be 1,100 tokens, and all token holders will have 10% more tokens in their
holding account.

Since interest is accrued continuously, however, just 1 second later, the total
supply has again increased. A 10% annualized rate gives us a continuous interest
rate of ~9.5% (`ln(1.1)`), which allows us to calculate the new instantaneous supply.

```
new_supply = previous_supply * exp(continuous_rate / seconds_per_year)
new_supply = 1100 * exp(0.095 / 31536000)
new_supply = 1100.0000033...
```

Since i-tokens are constantly accruing interest, they do not have the same
properties as normal tokens. Most importantly, i-tokens held in two different
accounts are only fungible if they have been updated to the same time. For example,
10 i-tokens in the year 1999 are worth much more than 10 i-tokens in the year 3000, since
those tokens from 1999 have over 1,000 years of interest (see Futurama). Because
of this, we have to compound interest before any minting, burning, and

> transferring.

Reviewer: Good general introduction with accessible background and problem statement. :)

### Initial Attempt

The concept seems simple to implement. We can create an SPL token, allow
tokens to be minted by a new "interest accruing" program, and then make sure to
call it regularly.

This solution has a few problems. First, how do we compound interest regularly?
We can create a separate service which periodically sends transactions to update
all of the accounts. Unfortunately, this will cost the service a lot, especially if
there are thousands of accounts to update.

We can enforce compounding before any transfer / mint / burn, but then all programs
and web consumers need to change to make these update calls. How do we change programs
and wallets to properly update accounts without creating a big `if is_i_token` case
all over the code?

This solution creates friction for almost all SPL token users: on-chain programs,
wallets, applications, and end-users.

> ## Proposed Solution

Reviewer: a bit lazy question, but how our https://github.com/solana-labs/solana-program-library/tree/90b4b33775b0a464ca443963e65ebc80756a3e2e/token-lending is working today to pay interests from lending without this i-token is still in proposal stage?

Author: No worries!  In its current form, the liquidity providers for token lending get yield directly from the borrowers in the form of repayments.  So if you deposit USDC into the liquidity pool, and no one borrows USDC, then you don't get anything.  Maybe we should think about that as more of a liquidity provider fee?

On the other hand, the i-token generates a constant interest like a savings account with a bank, so you always get a fixed percentage gain.

worth 100 tokens. If I transfer 1,000 tokens today, the program converts that and moves
10 shares. One year later, 1 share is worth 120 tokens, so if I transfer
1,000 tokens, the program converts and only moves 8.33 shares. The same concept
applies to mints and burns.

Following this model, we never have to worry about compounding interest. We
simply need to call the right program to perform instructions.

## Required Changes

A new interest-bearing token program is only one component of the overall solution,
since i-tokens need to be easily integrated in on-chain programs, web apps,
and wallets.

### Token Program Conformance

Since our solution entails the creation of a new token program, we're opening the
floodgates for other app developers to do the same. To ensure safe and easy integration
of new token programs, we need a set of conformance tests based on the current
SPL token test suite.

A new token program must include all of the same instructions as SPL token,
but the "unchecked" variants should return errors. Concretely, this means the
program must implement:

* `InitializeMint`
* `InitializeAccount`

> * `InitializeMultisig` -- TODO deprecate this and use a more general multisig?

Reviewer 2: Yes, I think this instruction should be deprecated

* `Revoke`
* `SetAuthority`
* `CloseAccount`
* `FreezeAccount`
* `ThawAccount`

> * `TransferChecked`

Reviewer: This new `TransferChecked` accepts amount in terms of token amount from `GetHoldingBalance`, not in terms of share amount?

If the amount is continuously increasing indeed, I think we need some magic number recognition (`-1`) to move all of amount out of the given holding account to `CloseAccount`.

Author: Yes, the idea is that all interactions from the outside are in terms of tokens, and not shares.

That's a good point regarding draining everything out of an account.  Are you suggesting adding an optional holding account parameter to `CloseAccount` which moves whatever is left in the holding to another account?

Reviewer: yeah, some thing like that? If share's amount is changing at slot generality, i thought it might hard to transfer all of then at once by end users.

I like `TransferChecked(u64::max)` than `Close()` because it's more like `transfer` than `close`. at user side, such ALL transfer can always paired with `Close()` instruction.

Reviewer 3: > I like `TransferChecked(u64::max)` than `Close()` because it's more like `transfer` than `close`. at user side, such ALL transfer can always paired with `Close()` instruction.

I'm a little weary of setting the precedent of using sentinels like this, especially in an amount field.  A `TransferAllChecked(mint)` instruction would work just as well and share most logic with `TransferChecked(amount,mint)`

Reviewer: Yeah, I like `TransferAllChecked(mint)` :)

Reviewer: @joncinque Could you add this new `TransferCheckedAll` or `TransferAllChecked` into the text? so that we can mark this thread as resolved?

Author: Yes, thank you for your feedback!

* `ApproveChecked`
* `MintToChecked`
* `BurnChecked`
* `InitializeAccount2`

And the program should throw errors for:

> 

Reviewer 2: ```suggestion

* `InitializeMultisig`
```

There are also new read-only instructions, which write data to a provided account
buffer:

* `GetHoldingBalance`: balance of the holding
* `GetHoldingOwner`: owner's public key
* `GetMintSupply`: mint supply
* `GetMintAuthority`: mint authority's public key
* `GetMintDecimals`: mint decimals

See the Runtime section for more information on how these are used.

Programs and wallets that wish to support i-tokens must update the instructions
to use all `Checked` variants and use the new read-only instructions to fetch
information.

The SPL token library provides wrappers compatible with any conforming program.
For the base SPL token program, the wrapper simply deserializes the account data
and returns the relevant data. For other programs, it calls the appropriate
read-only instruction using an ephemeral account input.

The structure for new holding and mint accounts must follow the layout of the
existing SPL token accounts to ensure compatibility in RPC when fetching pre and
post token balances.

For a holding, the first 165 bytes must contain the same information
as a normal SPL token holding. The following byte must be the type of

> account (ie. `Holding`), and after that, any data is allowed as required by the

Reviewer: (off-topic: I want the generic realloc feature also...)

new token program.

> The same applies for mints, but only for the first 82 bytes.

Reviewer: Cool, so it's finally possible to embed arbitrary metadata for NFTs?

Author: Exactly!  The custom token program could even do things like track who owned the NFT, restrict transfers, etc

The `amount` field in an SPL token holding should contain a balance in order

> to properly integrate with RPC's `preTokenBalances` and `postTokenBalances`.

Reviewer: `amount` here as in `share`'s amount? Ultimately, `{pre,post}TokenBalances` holds enough information to reconstruct the delta of balances as in `token`'s amount at the time?

Author: I'll clarify this part too -- `amount` would actually be the token amount in this case, and not the `share` amount.  Otherwise, I think RPC would need to perform a calculation in order to convert to "UI Amount".  Unless, are you suggesting that the read-only `GetBalance` instruction is run here as well?

Reviewer: > are you suggesting that the read-only GetBalance instruction is run here as well

Yeah, I was implying that while worrying about the computing cost. Otherwise, the method would be of less merit... That's because RPC consumers always call another call to rpc? I don't think they can tolerate to display share's amount (internal thing) to end users.

Author: Yeah, that's fair.  I think we should keep caching the token amount somewhere as an implementation detail.  This might also help out with ledger support, but I'll discuss that with @t-nelson

Reviewer 2: I think `amount` needs to be shares instead of balance, because all token instructions still take shares as input.  The `uiAmountString` should be the current real balance though

Reviewer 3: > Yeah, that's fair. I think we should keep caching the token amount somewhere as an implementation detail. This might also help out with ledger support, but I'll discuss that with @t-nelson

I wouldn't do any extra work account-side the the Leger app.  All it gets is a serialized transaction message, no other supplemental info like this.

Reviewer: > 

Hmm, I thought `amount` of all token/i-token instructions are `balance-amount`/`token-amount`, not `share-amount`. That's because it would be easier for consumers of i-token (end-user, devs) to use in that way?

Say, defi apps want to do `do something() if i-token is more than NN <TICKER>`. Then, if the instruction takes in terms of shares. we need to provide to convert back the value returned from getHoldingBalance into share's value?

I think atoken takes amount in token unit, not share unit:

https://github.com/aave/protocol-v2/blob/3d3dfc01bad46a1b4df6f4bdf47fd8845df848b0/contracts/protocol/tokenization/AToken.sol#L373-L384

Reviewer 2: I was imagining `getHoldingBalance` would return the token unit only.  To get the share unit, one decodes the token holding account like what's done today to read the SPL Token account balance

Reviewer 4: It's not clear to me whether you are all talking about the same `amount`: the `amount` stored within the first 165 token-holding state? or `amount` passed in an instruction/returned from a read-only instruction request?
Based on my understanding of the proposal so far, I assumed the former would be in terms of shares, the latter in terms of tokens.

> are you suggesting that the read-only `GetBalance` instruction is run here as well?

Seems to me like this is necessary. I am also envisioning what mvines just described ^

Author: I added a "Proposed Layout" section that might help clarify where my thoughts are on this, and clarified "token amount" vs "share amount".

### Token Program Registry

We need the [token-list](https://github.com/solana-labs/token-list) to include
vetted SPL token-conforming programs.

TODO consider adding Rust and C versions of the registry for on-chain programs and RPC

> TODO consider detailing the process of including a new program

Reviewer: I think it's also desirable to register group of token mints by its program_id.

Author: Ah yes, good point

### Runtime

> #### Ephemeral Accounts

Reviewer: maybe like this?: https://github.com/solana-labs/solana-program-library/tree/master/shared-memory

Author: Yes, the concept is based on the shared memory program.  I'm worried that the shared memory program adds a lot of friction for on-chain programs though -- it means that every instruction that uses tokens must *also* pass in a shared memory account and the shared memory program.  To make things easier, let's add it directly into the runtime.  What do you think?

Reviewer: yeah, you have a point. and indeed you have a good point for the friction of future upgradability of underlying programs.

Maybe, that's another selling point of my idea of dynamic AccountInfo construction via preflight? :)

Reviewer: That means even if we add some nice run-time-sanctioned special program, existing programs might not be able to use it for the possibility of breaking the AccountInfo list compatibility.

In order to reduce friction when using read-only instructions on other programs,
we need the ability for on-chain programs to dynamically create "ephemeral
accounts", which only contain data, can be passed through CPI to other programs,
and disappear after the instruction.

For a program using i-tokens, to get the balance of a holding, the program creates
an ephemeral account with 8 bytes of data, passes it to the i-token program's
`GetHoldingBalance` instruction, then reads the information back as a `u64`. At the
end of the instruction, the account disappears.

TODO What's the best approach to implement this?  For example, does an ephemeral

> account even need a public key?

Reviewer 2: This is probably just a syscall, `create_ephemeral_account(program_id, size)`, that returns an `AccountInfo`.  No public key required, the lamport balance is 0 (and probably the instruction fails if the lamport balance is !0 when the current instruction exits).

Flesh out more details with @jackcmay  though

Reviewer 5: Yeah, that's what I had in mind, could add information about the granted ephemeral accounts to `InvokeContext` so we can track it between calls.  Need to flesh out granting memory access to the program, etc...

This solution avoids the need to pass additional scratch accounts to almost all
programs that use tokens.

TODO what limits do we consider for these? A certain number of accounts?
A total amount of bytes? Regarding security, this breaks the previous

> check that a CPI must use a subset of accounts provided to the calling program.

Reviewer: my another wildest idea is separation of account data views for internal and external at the invoke context boundary:

_account.data translation between on-chain data and {in-transit/in-tx} data_

In other words, we want to _redefine_ the current spl-token account.data layout is the ABI itself.

1. If i-token holding account is referenced from non-owned program, it first converted from share's amount to token's amount by runtime by triggering specialized instruction in the i-token program (like specifying Trans1ate11111) and passes the translated account.data to the invoking program. This is on-chain => in-transit translation.

2. If the invoking program internally invokes i-token instruction with the i-token account, it's converted back to share's amount. This is in-transit => on-chain translation. The i-token program only deals with share's amount, internally.

3. Finally, after exiting from the invoked i-token program, the invoking program sees another translated view of mutated token accounts after yet another translation.

pro: external programs work as before without `GetHoldingBalance` and its friends. This means there will be single api for both bare tokens and i-tokens. Otherwise, generic token handling program code always must handle the 2 similar api for the two.
con: might be confusing. more tx processing load due to extra translations?

my observation is that tokens are special beast in the blockchain ecosystem: We want it to behave similarly across its breeds, while we want to customize it as much as possible for the added value at the same time. So, let's take this property to the extreme: minimally acceptable compatible interface is the abi itself of the bare token data.

This is like virtualization and duck-typing a bit.

Author: This is a really good idea too!  It would be extremely useful to have a unifying representation that can be used across programs, and doing it through these translations is a slick way of re-using code, eliminating the `Get*` instructions, and eliminating all friction for existing token users.  This was actually one of my first thoughts, but I was worried it would be confusing for app developers (as you point out), and instead opted for the simplest possible `Get*` instructions.

Let's see what other people think about it -- I like it! :smile:

Reviewer 2: What's stored in the "standard" part of the i-token holder account data between transactions in this case?  If not the share amount then wouldn't it be the balance at the time of last instruction executed, which could now be wrong/misleading to the user

Reviewer 5: Is the runtime doing the translation, if so how does it know if it should be doing the translation or not?  Does the runtime have to be aware of which accounts are i-token and which are not?

Reviewer: @mvines 

> What's stored in the "standard" part of the i-token holder account data between transactions in this case?

I meant the "standard" part of the i-token holder account will be in token's amount, not share's amount. Overall of my review comments, I wanted to hide the share's amount thing as much as possible, like ERC20. I thought ERC20's approch is right because exposing share's amount outside its own program might too confusing.

>  If not the share amount then wouldn't it be the balance at the time of last instruction executed, which could now be wrong/misleading to the user

Between transactions, I think this is fine. After all, i-tokens's balance is expected to continuously increasing? One concern is to transfer all of it (thus, I proposed `transferCheckedAll`) I think aTokens are seeing similar situation.

Also, as for between instructions and cpi invocations inside a single TX, i-token's token & share amount shouldn't change unless mutated so. In other words, time must not progress inside a tx. I think this is safe assumption, considering atomicity and ACID guarantee.

Reviewer: > Is the runtime doing the translation, if so how does it know if it should be doing the translation or not? Does the runtime have to be aware of which accounts are i-token and which are not?

@jackcmay that's good questions. I think there are several options to indicate the flag like this:

- Make runtime recognize wrapper Loader address TranslatedBPFLoader111. (bad orthogonality though, for upgradable/non-upgradable)
- Sneak some flag into the elf header?
- or some kind of marker symbol in the elf symbol table (or even the entrypoint of the translation routine)?

Reviewer 2: > After all, i-tokens's balance is expected to continuously increasing?

It could decrease :)
Imagine an iToken that self-destructs after N epochs, or loses value on each transfer.  An NTF iToken might be designed to work this way.   Your share remains the same but the balance decreases.

#### Dynamic sysvars

The i-token program also requires the use of dynamic sysvars. For example,
during the `GetHoldingBalance` instruction, the program needs to know the current
time in order to properly convert from token amount to share amount.

Without dynamic sysvars, we need to create a system to tell what sysvars are
needed for different instructions on each program, which creates complications

> for all users.

Reviewer: This is my very wild idea but, I'm thinking it's about time for us to realize some dynamic programmable `AccountInfo`s creation for composable TXes:

_use preflight transaction execution for dynamic AccountInfo generation_

So, when tx is executed by preflight in the rpc, it records actually dynamically referenced accounts and returns the list to the client. Then, client resends the real tx to the rpc with added AccountInfos. 

```rust
fn entrypoint(account_infos: DynamicAccountInfos) {
  account_info.get(pubkey, is_readonly, is_signer)?
  let inner_account_info = account_info.create_inner_account_info();
  inner_account_info.push(...);
  ...
  invoke(ix, &inner_account_info);
}

impl DynamicAccountInfos {
  // in pre-flight simulation execution environment, record requested account keys into HashMap and modes then actually return the account data via syscall
  // in real execution environment, return the account if tx referenced it correctly; otherwise return Err
  fn get(...) -> Result<AccountInfo> {
  }
  // create AccountInfo with empty positional accounts, while cloning the HashMap for the dynamically-referenced accounts; this is like ARGV/Env relation.
  fn create_inner_account_info()
  }
}

```

pro: downstream developer of any defi app doesn't need care about which accounts to pass, minimum friction for writing newer code and maximum compatibility is retained, because pre-flight is done off-chain, relatively less load on the on-chain processing. Also, this enables vectorization of write-referenced accounts. Lastly, greatly flexibility at the hands of program devs
con: 1 extra round trip between client and rpc. Offline tx is a bit more tedious. This can be abused?

Or if we allow delaying and buffering the tx internally inside bank (step 1: pre-flight, step 2: actual execution):

pro: no round-trip at the client side.
con: harder proper cost structure, 2x load on banking/replaying path (also optimistic locking can be employed to minimize the load?)

Reviewer: Putting aside customized ntfs, my motivating use case was erc20's paxg, which collects its special small fees everytime its token is transferred.

Author: This is quite an idea!  We definitely need some model for declaring the required accounts in an instruction.  This is a cool solution for accounts where the pubkey is known.  That covers sysvars, programs, global state accounts, dependent accounts (e.g. the mint associated with a token swap), but would not cover certain positional arguments, correct? How would this work with positional arguments where we don't know the pubkey?

Reviewer: > would not cover certain positional arguments, correct? How would this work with positional arguments where we don't know the pubkey?

Seems my idea is grokked by you, correctly. this is a good question. :) Well, this is possible. thanks to the force of blockchain's deterministic behavior.
The reason is 2 phased:

Firstly, it's always determinable to dynamically know the actual needed pubkeys at the time of simulation from runtime perspective. That's because conceptually, tx must have been constructed with priori knowledge of correct understanding for positional parameters.

So, consider like this (maybe you had this code in your mind?):

```
     let foo_bar = FooBar::unpack(account_info.next());
    foo_bar.do_baz();
```

Given this code, the right pubkey for the positional account can't be known by runtime simulation. But this is fine. The caller (another cpi-ed program or ultimate defi web browser app) must know how to explicitly specify the right pubkey. Any semantically wrong pubkey will lead to `Err`. Conversely, if the program did work by passing arbitrary account at the position, then the account is extra (unneeded) positional argument or just the program is badly written.

Secondary, let's think about this from theoretical stand point. There's two key roles here. program author and tx creator. Ultimately, both parties just want to ensure expected behavior for actual state transitions in the form of `s' = program(s, tx) where s = list(account)`. program author want program to refer to accounts for its internal book-keeping but also parameterize some ones as input of the state transition _to be a actually useful program to be used by other parties_. In other words, the parametarized accounts are unknownable up to the actual execution. So, at this moment, list(account) could contain unknown elements. Then, tx creator alike want the program to refer to accounts for their utility using their knowledge of expected behavior of the program (Also don't want to know which internal account is needed for the program's book-keeping!). At this moment, the list(account) should be sufficient set of accounts to realize the deterministic expected behavior, containing no undetermined account. Otherwise, program author or tx creator did wrong.

In other words, we can categorize AccountInfos into two kinds: static account info, dynamic account info:
static account info: determined by program author; ultimately, all can be determined by runtime simulation.
dynamic account info: determined by tx creator with their knowledge.

Also, the good thing for the dynamic account list generation is that, it's not actually security concern to allow the program author to add arbitrary number of account info to the tx thanks to our signer bit and strict account ownership runtime mechanism. After all, all program can read ( but not write) any state of accounts as before. :)

Reviewer 2: I think this is out of scope for this proposal, which is already quite large and highly impactful on the ecosystem.  Overhauling how accounts work should be separate and if done before i-tokens could be used here as well.

Author: I see what you mean here.  I was worried more specifically about dynamic *nested* accounts.  For example, in the case of token lending, you need the lending reserve as a dynamic input, but inside of that are all of the extra accounts *inside* the reserve, like the token mint for that reserve.  We can say that a tx creator should just know all of that up front, but since we're talking about a model that declares everything that you'll need up front, it would be extremely useful to fetch dependent accounts as well.  I think your model could possibly be extended to make this work though :-)

Either way, since this is out of scope for the current proposal, I will resolve the comment.  Thanks a bunch!

Reviewer: > I was worried more specifically about dynamic nested accounts. For example, in the case of token lending, you need the lending reserve as a dynamic input, but inside of that are all of the extra accounts inside the reserve, like the token mint for that reserve. ... I think your model could possibly be extended to make this work though

@joncinque I think the current model actually works for dynamic nested accounts because the tx simulation also scans any nested CPI for on-demand account scanning and those accounts are implicitly passed to the nested InvokeContext like environment variables. This being said, existing programs are need to be modified to tell the address hint to the runtime at least. Also, I'll try to slip this facility in the timeline of itoken transition across the board.

Thanks to #16218, this will be kind of easy, I hope.

Author: @ryoqun this all depends on dynamic account loading in the simulation mode, correct?  It looks like we do have a model tht works then :smile:

Reviewer: wow, 2 months has passed.. it've got some time to write this finally: https://github.com/solana-labs/solana/pull/17796

### RPC

The validator's RPC server needs to properly handle new token program accounts

> for read-only instructions and pre / post balances.

Reviewer 2: Probably worth calling out how RPC will learn about the known program ids.   I feel like this could be a file we host somewhere, like here: https://github.com/solana-labs/token-list/releases/latest,  and RPC downloads it automatically at startup

#### `getBalance` and `getSupply`

Instead of deserializing SPL token accounts and returning the data, the
RPC server runs read-only transactions on the i-token program.

RPC simply uses the same flow as on-chain programs: create an ephemeral account
then runs the read-only transaction onto the bank requested.

> TODO how do we set the cost of read-only transactions? Do we lower it specially

Reviewer: Maybe, this could just be free? I think we're already doing kind of read-only transactions via the pre-flight mechanism.

Author: Sorry, this was unclear from my side.  I was referencing compute cost here.  The transaction should definitely be free, just like a pre-flight transaction.  I was wondering if we should reduce the compute cost from 200k to something lower, since read-only instructions should be faster.

Reviewer: I see. thanks for clarification. I think this can be undecided for now. More proper general fee structure will be rolled out like tx-wide compute budget (cc: @jackcmay )

for read-only RPC calls? The i-token program will probably be one of the more
cost-intensive computations possible since it is performing present value
discounting.

#### `preTokenBalances` and `postTokenBalances`

As mentioned earlier in the Token Program Conformance section, the i-token
program caches balances in the `amount` field of the SPL token holding. This
should only be used by RPC, and not by programs to get the balance of a non-SPL
holding.

### Associated Token Program

The Associated Token Program needs to support all token programs seamlessly.
Currently, it performs the following sequence:

* `transfer` required lamports
* `allocate` space
* `assign` to the SPL token program
* call `InitializeAccount` on SPL token

> This does not work for an opaque token program, because we do not know the size required

Author: Yep, definitely

> This does not work for an opaque token program, because we do not know the size required

Reviewer: fyi, mostly this is supported by the tip of the spl repo: https://github.com/solana-labs/solana-program-library/pull/1364/files#diff-327d4c08bad2c8914b9075c60d60d32d54c2f40d73850ee19854550a9490d1d1R33

just we need to deprecate program-sdk facing helper fn which doesn't take the program id.

from the outside. Conversely, if we allow a token program to take the lamports it wants,
a malicious token program could take too much from users.

> To get around this, the Token Program interface has a new `CreateHolding`

Reviewer: off-topic: but I want to apply these changes when we have a chance to side-load the associated token program: https://github.com/solana-labs/solana-program-library/issues/1371

Author: Oh yeah, that would make things a lot easier to use.  The instruction to the associated token program can definitely be included in the wrapper libraries.  Programs that want to opt in would need to pass the associated token program as well, which isn't a big thing to ask.

instruction, which only allocates space and assigns the account to the program.
Once the space is allocated, the Associated Token Program can transfer the

> required lamports, and return an error if that number is too large. The process

Reviewer: how much would be too large like over 10KB?

Reviewer: I think for the case of NTFs (i.e. somewhat controlled environment), it might be desirable to have an option of raising the bar to carry around somewhat rich metadata of ownership transitions.

Author: I'm not sure to be honest -- 10KB seems pretty big for an account, but that would be a safe maximum.  Perhaps we start smaller and grow as needed?

In the case of an NFT, they could put most of the information in the mint so that token accounts stay small.  The mint is passed for all transfers so the program can do any necessary checks there.

Reviewer 5: Might want to put the data into a separate account.  The NFT data might be large and we don't need to pass it around for transfers, etc.  Just put a reference to the NFT data account in the mint?

Reviewer 6: I agree with @jackcmay NFTs data should go to a separate metadata account, the same way as in ERC721Metadata

Reviewer: (this is out of scope of this proposal...)

@jackcmay @bartosz-lipinski Yeah, creating and separating account and referencing would work for normal NFT metadata. But, I was thinking about complex NFT which want to mutate the state of NFT metadata or want to hold past transfer history somehow. In that case, we need a writable flag for referenced mint account or additional NFT metadata, which isn't conforming to this i-token abi. So, I was thinking about putting those mutable NFT state in the already writable accounts (i.e. the dest token holding account with somewhat large buffer).

Then, ultimately I dreamed some fancy idea to conform the complex NFT to the i-token: https://github.com/solana-labs/solana/pull/15927#discussion_r595842064

becomes:

* call `CreateHolding` on the token program (allocate and assign)
* calculate rent requirement based on the data size of holding account

> * `transfer` required lamports

Reviewer 2: ```suggestion
* `SystemInstruction::Transfer` the required lamports to make the holding account rent-exempt
```

* call `InitializeAccount` on the token program

### Web3 / Wallets

To properly support i-tokens, wallets and applications will need to make the
following changes:

* get holdings: query for holdings owned by the public key, for all official

> token programs listed in the registry (not just `Tokenkeg...`)

Reviewer: this means `getTokenAccountsByOwner` needs smarter secondary indexes.

Author: Very true, I'll add that as a consideration

* get balances: avoid deserializing holding data, and instead use
`getBalance` from RPC
* transfer tokens: only use `ApproveChecked` and `TransferChecked` on the token
program
* create holding: avoid directly creating an account and initializing,
and use the associated token account program instead

> 

Reviewer: our of curiosity, I wonder how ledger support aTokens? Maybe, we should follow the precedent?

> 

Reviewer 4: Hmm, shouldn't the Ledger display token amount, not share?
Based on the description in the Proposed Solution section, I understood that the program would take care of all conversions to shares, thus the instruction data passed in (and displayed on the Ledger) would be in tokens.

> 

Reviewer 3: > our of curiosity, I wonder how ledger support aTokens? Maybe, we should follow the precedent?

AFAIK, it doesn't. Just standard ERC20

> 

Reviewer 2: > Hmm, shouldn't the Ledger display token amount, not share?
> Based on the description in the Proposed Solution section, I understood that the program would take care of all conversions to shares, thus the instruction data passed in (and displayed on the Ledger) would be in tokens.

All instructions process `shares` -- token amount is the `uiAmount`, which may now be a more involved computation than multiplying in the number of decimals places.
The Ledger only gets `shares` since it can't execute the code provided by the program to compute a `uiAmount` based on the current start of the chain

> 

Reviewer 4: I'm very confused, then. I thought converting back from tokens to current shares -- with whatever involved logic -- is what the program does when processing the instruction. The Ledger only has the instruction data to work with.

> Yes, the idea is that all interactions from the outside are in terms of tokens, and not shares.

Ref Jon's comment: https://github.com/solana-labs/solana/pull/15927/files#r595974135

although the token holding amounts are constantly changing, each token holder's
proportion of total supply stays the same. Therefore, we can satisfy the
fungibility requirement for tokens on proportional ownership. The actual token
amounts used for token transfers / mints / burns are just an interface, and
everything is converted to proportional shares inside the program.

Let's go through an example. Imagine we have an i-token mint with a total of
10,000 proportional shares, a 20% annual interest rate, and 1 share is currently
worth 100 tokens. If I transfer 1,000 tokens today, the program converts that and moves
10 shares. One year later, 1 share is worth 120 tokens, so if I transfer
1,000 tokens, the program converts and only moves 8.33 shares. The same concept
applies to mints and burns.

Following this model, we never have to worry about compounding interest. We
simply need to call the right program to perform instructions.

## Required Changes

A new interest-bearing token program is only one component of the overall solution,
since i-tokens need to be easily integrated in on-chain programs, web apps,
and wallets.

> ### Token Program Conformance

Reviewer: I think we can side-load the tokenkeg to support only these newly added instructions. So that newer downstream projects doesn't support the two instruction set standards indefinitely?

For that, I think we need to be careful not to collide with existing enum variant numbers. :)

Reviewer: if we could realize this, I think we have less motivation of these fancy thing: https://github.com/solana-labs/solana/pull/15927/files#r595931233

Author: I believe the idea is to side-load tokenkeg with the new instructions, any objections @mvines ?

Reviewer 2: Yep we'll need to side-load both tokenkeg and ATA to support iTokens

Since our solution entails the creation of a new token program, we're opening the
floodgates for other app developers to do the same. To ensure safe and easy integration
of new token programs, we need a set of conformance tests based on the current

> SPL token test suite.

Reviewer: Also, for further easiness of creating interesting i-tokens, I think we should provide some scaffolding mechanism by rust.

maybe like this?:

```
trait InterestBearingToken {
  fn token_amount_to_share_amount() -> u64,
  fn share_amount_to_token_amount() -> u64,
}

trait Token<T: InterestBearingToken> {
  fn process_transfer_checked(...) {
     ...
     self.process_transfer_checked_internal(T::token_to_share(amount));
     ...
  }
  fn process_get_holding_balance(...) {
    ....
     return T::share_to_token(self.amount)
  }
}
``` 

As a testbet, I think we can refactor current existing implementation like this?

The intention here is that avoiding the plain copy-paste; because this will be nightmere for patching.
So, itokens developer only need to care about their cool present-value/recoup/interest/etc quant calculation. :)
Then, pretty basic invariant of no phantom tokens and authorization-checking code (this is pretty battle-tested now) can be reused.

Author: Very good idea, we should absolutely re-use as much of the existing SPL token code as possible

A new token program must include all of the same instructions as SPL token,
but the "unchecked" variants should return errors. Concretely, this means the
program must implement:

* `InitializeMint`
* `InitializeHolding` (previously `InitializeAccount`)
* `Revoke`
* `SetAuthority`
* `CloseHolding` (previously `CloseAccount`)
* `FreezeHolding` (previously `FreezeAccount`)
* `ThawHolding` (previously `ThawAccount`)
* `TransferChecked`
* `ApproveChecked`
* `MintToChecked`
* `BurnChecked`

> * `InitializeHolding2` (previously `InitializeAccount2`)

Reviewer: ugh `2`? I really want to get rid of this. ;)

Author: `InitializeHoldingWithoutOwnerAccount`?

Reviewer 4: `InitializeHoldingWithOwnerAddress`?

And the program should throw errors for:

* `Transfer`
* `Approve`
* `MintTo`
* `Burn`
* `InitializeMultisig` (use a more general multisig)

New instructions required:

* `CreateHolding` (TODO name): only performs `Allocate` and `Assign` to self,
useful for creating a holding when you don't know the program that you're
interacting with. See the Associated Token Account section for how to use this.

There are also new read-only instructions, which write data to a provided account
buffer:

* `GetHoldingBalance`: balance of the holding
* `GetHoldingOwner`: owner's public key
* `GetMintSupply`: mint supply
* `GetMintAuthority`: mint authority's public key

> * `GetMintDecimals`: mint decimals

Reviewer 4: @joncinque can you tell me a little more about why HoldingOwner, MintAuthority, and MintDecimals are included in this list, and the other authorities and delegate are not?

Author: The idea was to limit the common interface as much as possible to the minimum needed for app developers, but the more I think about it, the more it seems we should not have these `Get*` instructions and instead have a `SerializeToSPL` instruction, which in practice works very similarly to the `Get*` instructions, but writes the whole holding or mint in the same exact way as the standard SPL form.  Meaning, you'd have any i-token holding account, not sure what it is, so you create an ephemeral account with 165 bytes, send it to the instruction, and then you can deserialize the ephemeral account into a `spl_token::Account`

Reviewer 4: I like `SerializeToSPL`, enabling program logic to apply to any part of the `spl_token::Account`/`spl_token::Mint`.

See the Runtime section for more information on how these are used.

Programs and wallets that wish to support i-tokens must update the instructions
to use all `Checked` variants and use the new read-only instructions to fetch
information.

The SPL token library provides wrappers compatible with any conforming program.
For the base SPL token program, the wrapper simply deserializes the account data
and returns the relevant data. For other programs, it calls the appropriate
read-only instruction using an ephemeral account input.

The structure for new holding and mint accounts must follow the layout of the
existing SPL token accounts to ensure compatibility in RPC when fetching pre and
post token balances.

For a holding, the first 165 bytes must contain the same information

> as a normal SPL token holding. The following byte must be the type of

Reviewer: is this spoofing-attack safe? What if a malicious token program populates its token holding account with the mint byte range being initialized with one of established spl-tokens?

Author: I want to be sure that I understand this -- you mean a spoofing attack where a malicious token program intentionally mixes up tokens and mints?  To me, that would mean the malicious token program doesn't work at all, not impacting other tokens.  How would it relate to the standard spl token in that case?

Reviewer: this is rather minor issue.

But, malicious program and its data accounts can minic as if it's a (hypothetical) well established tokens like USDZ living in the tokenkeg land. That's because malicious token program has completely control the bits.

Then, further assume that victim program is hard-coding mint address of USDZ but they're not checking account.owner/program_id as being tokenkeg. This  could trick the problem.

being `minor` is that because `TransferChecked` would fail anyway. But (unusual?) custody model might cause problem (by set_authority) because that might not write the faked account data unlike TransferChecked.

So, I think we should strongly recommend to check program_id everywhere and actually add these checks across the spl-token / i-token codebase.

So, they can even bypass the base spl-token deserialize check by adding 0 extra data intentionally. (So, at least 1 more byte worth nice security protection)

Also, `SerializeToSPL` could return whatever faked binaries.

Author: I see what you mean now.  That's definitely a consideration, and the best solution seems to force people to always check the program id of the account.  The new wrapper library will accept an `AccountInfo` for the token program, so we'll do the check there, which should protect most users.  There's still a risk of bad token programs being out there, which is why the registry will become an important component.

Reviewer 7: Agreed with @ryoqun that the possibility of mixing up mints and holding accounts is undesirable. I think the design should make it hard for token program designers to mess this up. Reserving the first byte for determining which account is being deserialized seems like the best approach (especially when custom token programs define other account types). Judging by account size was never meant to be a general approach.

account (ie. `Holding`), and after that, any data is allowed as required by the

> new token program.

Reviewer: Also, what if malicious program didn't add any extra data at all? Then, current HoldingAccount::unpack() returns Ok bypassing the account.data check. So, maybe, we should start to check account.owner as well when deserializing.

Author: Yes, definitely, the `HoldingAccount::unpack()` will only be called for accounts owned by `tokenkeg`

The same applies for mints, but only for the first 82 bytes.

The `amount` field in an SPL token holding should contain a balance in order
to properly integrate with RPC's `preTokenBalances` and `postTokenBalances`.

For i-tokens, `amount` will be in terms of tokens, and not shares, so that UIs
can properly display amounts moved.

### Token Program Registry

We need the [token-list](https://github.com/solana-labs/token-list) to include
vetted SPL token-conforming programs and group known mints by their program id.

> The token list will publish a file of known program ids at

Reviewer 3: This will be the crux of the Ledger app implementation.  Instruction parsing is currently gated on program id and flash storage is limited

Reviewer 2: As MVP, if the Ledger app doesn't recognize the program id (Tokenkeg) then could we display the program id to the user as well for confirmation?

Reviewer 3: Currently it's just going to bail on an unrecognized instruction and the user would have to blind sign.  Without a known program ID there's no way to route to the appropriate instruction parser, so there's nothing else to display the program_id with anyway.  We'll fail similarly on any new instructions in the alternate implementation

I've been debating trying to add a fallback mode that displays instruction info one at a time instead of collecting all of the info up front and coalescing them into a nice display for the whole transaction. I'm not entirely sure that will fit on the device alongside the current implementation though, so it may be one or the other.  I'm also somewhat dubious of this route in that the number of fields the user might be presented with could defeat the purpose of visual confirmation through monotony or making them race the blockhash expiration

Another option would be to move the SPL Token features into a separate Leger app where we can make more assumptions about what it's getting and/or make it easily extensible so projects can just fork it like KIN did the current app.  I'm not 100% sure this is viable though as LedgerHQ may disallow multiple apps using the same derivation path, which would effectively silo them.  I'll ask upstream

Author: @t-nelson how does `7320a19` look per our Ledger discussions?  I unfortunately couldn't remember what we decided on the associated token account, so I left a TODO there -- sorry about that!

* get balances: avoid deserializing holding data, and instead use
`getBalance` from RPC
* transfer tokens: only use `ApproveChecked` and `TransferChecked` on the token
program
* create holding: avoid directly creating an account and initializing,
and use the associated token account program instead

### On-chain Programs

To properly support i-tokens on-chain, programs will need to make these changes:

* multiple token programs: all programs that support SPL Tokens must always
include the program id of the token(s) holdings they're passing in.
For example, token-swap needs to accept separate token programs for token A and B
* hard-coded token program: avoid using `Tokenkeg...` directly, and delete `id()`
the SPL token code
* get balance: instead of deserializing the account data into an SPL holding, use
the new wrapper to deserialize or call `GetHoldingBalance` instruction with the
appropriate token program
* get owner: same as above, but using the `GetHoldingOwner`
* get supply: same as above, but using the `GetMintSupply`
* get mint authority: same as above, but using the `GetMintAuthority`
* get mint decimals: same as above, but using the `GetMintDecimals`

> * transfer: only use `TransferChecked`, which requires passing mints

Reviewer: considering https://github.com/solana-labs/solana/pull/15927/files#r597540110, I think any programs hard-coding some notable spl-tokens should start to check the authentity of given holding account by checking (program_address, mint_address) tuple, not only mint_address.

Author: Absolutely, we've been trying to push people in that direction as much as possible, but having it done automatically in the code will be even better

## Other New Token Programs

For awhile now, people have been coming up with ideas that likely require a new
token program to implement. These include:

> ### Voting Tokens

Reviewer: how about adding stake-pool's tokens with regard to its rewards?

Author: I'm not sure if we need a new token program for that -- an oracle on the stake pool seems like a better approach to me, but I might be missing something.  How would this work?

Reviewer: never mind, then. ;) I'm basically noob about the stake pool. I felt stake-rewards are like compounding-interest-link thing so, I wondered stake-pool _tokens_ could be nicely modelled around this i-token.

but the "unchecked" variants should return errors. Concretely, this means the
program must implement:

* `InitializeMint`
* `InitializeHolding` (previously `InitializeAccount`)
* `Revoke`
* `SetAuthority`
* `CloseHolding` (previously `CloseAccount`)
* `FreezeHolding` (previously `FreezeAccount`)
* `ThawHolding` (previously `ThawAccount`)
* `TransferChecked`
* `ApproveChecked`
* `MintToChecked`
* `BurnChecked`
* `InitializeHoldingWithOwnerAddress` (previously `InitializeAccount2`)

And the program should throw errors for:

* `Transfer`
* `Approve`
* `MintTo`
* `Burn`
* `InitializeMultisig` (use a more general multisig)

New instructions required:

* `CreateHolding` (TODO name): only performs `Allocate` and `Assign` to self,
useful for creating a holding when you don't know the program that you're
interacting with. See the Associated Token Account section for how to use this.
* `TransferAllChecked`: transfers everything from the source holding to the
destination holding. Useful for closing a interest-bearing token holding since the
balance is constantly changing.

There are also new read-only instructions, which write data to a provided account
buffer:

> * `SerializeHoldingToSPL`: write holding data in SPL token format

Reviewer: Hmm, I rather want to have better name. But there is no alternative name come to mind..  ;)

Reviewer 2: `NormalizeHolding`?

Author: The SPL part is important, so I propose `NormalizeToSPLHolding`

Reviewer 2: The SPL part is misleading.  `NormalizeToSolanaProgramLibraryHolding` doesn't really make sense.
SPL is not the same as SPL Token

Reviewer 4: `NormalizeToTokenkeg`? jk 😆 
I suppose to be completely explicit, it should be `NormalizeToSPLTokenHolding`, but imo `NormalizeHolding` is sufficient

Author: I stand overruled :smile: will do `NormalizeHolding`

For example, token-swap needs to accept separate token programs for token A and B
* hard-coded token program: avoid using `Tokenkeg...` directly, and delete `id()`
the SPL token code
* get holding data: instead of deserializing the account data into an SPL holding, use
the new wrapper to directly deserialize or call the `SerializeHoldingToSPL`
instruction with the appropriate token program
* get mint data: same as above, but using the `SerializeMintToSPL`
* transfer: only use `TransferChecked`, which requires passing mints

## Other New Token Programs

For awhile now, people have been coming up with ideas that likely require a new
token program to implement. These include:

### Voting Tokens

Outlined in a [GitHub issue](https://github.com/solana-labs/solana-program-library/issues/131),
voting tokens use additional data about how long a token has been held.

### Non-Fungible Tokens

It will be possible to add a lot of metadata regarding owners and traders, either
on a new mint or NFT account.

### Allow-list Tokens

Only a certain list keys are allowed (or not) to transfer to tokens, controlled
by the mint.

## Terminology

> * token amount: the amount of tokens as seen from the outside. For example, an

Reviewer: how about adding short entries for base spl-token and i-token?

destination holding. Useful for closing a interest-bearing token holding since the
balance is constantly changing.

There are also new read-only instructions, which write data to a provided account
buffer:

* `GetHoldingBalance`: balance of the holding
* `GetHoldingOwner`: owner's public key
* `GetMintSupply`: mint supply
* `GetMintAuthority`: mint authority's public key
* `GetMintDecimals`: mint decimals

See the Runtime section for more information on how these are used.

Programs and wallets that wish to support i-tokens must update the instructions
to use all `Checked` variants and use the new read-only instructions to fetch
information.

The SPL token library provides wrappers compatible with any conforming program.
For the base SPL token program, the wrapper simply deserializes the account data
and returns the relevant data. For other programs, it calls the appropriate
read-only instruction using an ephemeral account input.

The structure for new holding and mint accounts must follow the layout of the
existing SPL token accounts to ensure compatibility in RPC when fetching pre and
post token balances.

> For a holding, the first 165 bytes must contain the same information

Reviewer 4: Do we actually need to require that i-token holdings conform to any particular layout? As long as the program implements a `SerializeToSPL` method that is called anytime a caller wants to access the account as an SPL token, it seems like the underlying i-token holding could be whatever state is most efficient for the particular program.

For instance, in the specific case of interest-bearing tokens laid out below, there wouldn't be any need to store both share amount and a token amount (that may always be wrong). It could be instead:

```rust
pub struct Holding {
    // SPL Holding Fields
    pub mint: Pubkey,
    pub owner: Pubkey,
    pub share_amount: u64,
    pub delegate: COption<Pubkey>,
    pub state: AccountState,
    pub is_native: COption<u64>,
    pub delegated_amount: u64,
    pub close_authority: COption<Pubkey>,
}
```

where `SerializeToSPL` converts share_amount to current token_amount

Reviewer 2: We'll need to expose SerializeToSPL over RPC as well for `spl-token` cli and other front-ends to use if we do this, not bad though.

I guess the counter is that we generally expect iTokens to fully implement SPL Token (and then add some), so why wouldn't they need all these fields.  Implementation wise I think we want to encourage a common SPL Token library crate that is built on rather than everybody forking from day 1, so we can easily publish updates for security fixes.  Not having a common `struct` to rally around makes integration of new SPL Token releases much harder for the downstream

Reviewer 4: > We'll need to expose SerializeToSPL over RPC as well for `spl-token` cli and other front-ends to use if we do this, not bad though.

I think that we're going to need to do this regardless in order to return correct token amounts at any given time.

I see your point about new token programs being additive.
I suppose it is only 8 bytes. Just seems a waste, if amount is essentially going to become `unused`.

Author: I've updated the docs to enforce that new token programs conform their structs to look like SPL token, along with the important difference between UI amount and amount.

Without dynamic sysvars, we need to create a system to tell what sysvars are
needed for different instructions on each program, which creates complications
for all users.

### RPC

The validator's RPC server needs to properly handle new token program accounts
for read-only instructions and pre / post balances.

#### `getBalance` and `getSupply`

Instead of deserializing SPL token accounts and returning the data, the
RPC server runs read-only transactions on the i-token program.

RPC simply uses the same flow as on-chain programs: create an ephemeral account
then runs the read-only transaction onto the bank requested.

TODO how do we set the cost of read-only transactions? Do we lower it specially
for read-only RPC calls? The i-token program will probably be one of the more
cost-intensive computations possible since it is performing present value
discounting.

#### `preTokenBalances` and `postTokenBalances`

As mentioned earlier in the Token Program Conformance section, the i-token
program caches token balances (not share amounts) in the `amount` field of the
SPL token holding. This should only be used by RPC, and not by programs to get

> the balance of a non-SPL holding.

Reviewer 4: I don't think we should use this cached value it all. It might be out of date for preTokenBalances, and I don't think we can even guarantee its correctness for postTokenBalances unless we require every token account be loaded as writable (and rewritten) every transaction.

Author: The idea was to require programs to rewrite it on all transactions, since RPC will need access to it.  Do you think pre / post tokenBalances could call SerializeToSPL every time?  I'm not familiar with the flow there

Author: We discussed this last week, and we can have RPC use `SerializeToSPL` to fetch the correct balances, which eliminates the requirement to have the same layout

voting tokens use additional data about how long a token has been held.

### Non-Fungible Tokens

It will be possible to add a lot of metadata regarding owners and traders, either
on a new mint or NFT account.

### Allow-list Tokens

Only a certain list keys are allowed (or not) to transfer to tokens, controlled
by the mint.

## Terminology

* token amount: the amount of tokens as seen from the outside. For example, an
interest-bearing token holding with 10 in token amount is converted to a proportional
ownership of 0.5% of the total supply. This amount is always shown in UIs, and
also cached in the `amount` field of 

* share amount: the proportional ownership of a holding in terms of the total
supply of shares. Internally, the interest-bearing token program uses shares to
mint / transfer / burn, and this amount is never shown in UIs.

## Proposed Layout

### Mint

```rust
pub struct Mint {
    // SPL Mint Fields
    pub mint_authority: COption<Pubkey>,
    pub supply: u64, // Cached token amount of current supply, used by RPC
    pub decimals: u8,
    pub is_initialized: bool,
    pub freeze_authority: COption<Pubkey>,

    // Interest-bearing Mint Fields

>     pub share_supply: f64, // The total amount of shares outstanding

Reviewer 4: Hopefully we won't actually use f64 for anything, though ;)

Author: From a conversation with Jack, who was talking about how well they work in the vm :-)

SPL token test suite.

A new token program must include all of the same instructions as SPL token,
but the "unchecked" variants should return errors. Concretely, this means the
program must implement:

* `InitializeMint`
* `InitializeHolding` (previously `InitializeAccount`)
* `Revoke`
* `SetAuthority`
* `CloseHolding` (previously `CloseAccount`)
* `FreezeHolding` (previously `FreezeAccount`)
* `ThawHolding` (previously `ThawAccount`)
* `TransferChecked`
* `ApproveChecked`
* `MintToChecked`
* `BurnChecked`
* `InitializeHoldingWithOwnerAddress` (previously `InitializeAccount2`)

And the program should throw errors for:

* `Transfer`
* `Approve`
* `MintTo`
* `Burn`
* `InitializeMultisig` (use a more general multisig)

New instructions required:

* `CreateHolding`: only performs `Allocate` and `Assign` to self,
useful for creating a holding when you don't know the program that you're
interacting with. See the Associated Token Account section for how to use this.
* `TransferAllChecked`: transfers everything from the source holding to the
destination holding. Useful for closing a interest-bearing token holding since the

> balance is constantly changing.

Author: Correct, we transfer shares, which means we no longer have a need for this.  If it seems nice to have, I'm happy to keep it in!

information.

The SPL token library provides wrappers compatible with any conforming program.
For the base SPL token program, the wrapper simply deserializes the account data
and returns the relevant data. For other programs, it calls the appropriate
read-only instruction using an ephemeral account input.

For interest-bearing tokens, after `NormalizeHolding` and deserialization
into a `spl_token::Holding`, the `amount` field will be in terms of tokens, and
not shares, so that UIs can properly display amounts in terms of tokens.

### Token Program Registry

We need the [token-list](https://github.com/solana-labs/token-list) to include
vetted SPL token-conforming programs and group known mints by their program id.

The token list will publish a file of known program ids at
[token-list releases](https://github.com/solana-labs/token-list/releases/latest),
to be used by RPC and programs.

The entries in the token-list are signed using a managed key. As wallets or
Ledgers are asked to send transactions using a particular token program, they can
cross-check the mint or program information using a signature check on the 
token-list data.

> TODO consider adding Rust and C versions of the registry for on-chain programs,

Reviewer 3: I think it will be sufficient to add support to our _client_ libraries for fetching, caching and querying the token-list. God help us if that ever includes a C implementation :sweat_smile:

Author: right, i'll clarify the client library bit.  and good point about C integration :laughing:

### On-chain Programs

To properly support i-tokens on-chain, programs will need to make these changes:

* multiple token programs: all programs that support SPL Tokens must always
include the program id of the token(s) holdings they're passing in.
For example, token-swap needs to accept separate token programs for token A and B
* hard-coded token program: avoid using `Tokenkeg...` directly, and delete `id()`
the SPL token code
* get holding data: instead of deserializing the account data into an SPL holding, use
the new wrapper to directly deserialize or call the `NormalizeHolding`
instruction with the appropriate token program
* get mint data: same as above, but using the `NormalizeMint`
* transfer: only use `TransferChecked`, which requires passing mints

### Ledger

Ledger integrates with token-list in order to provide verified token program ids
and token tickers to the Ledger SDK and eliminate the need to store static
information in the Ledger Solana App.

A wallet or CLI with Ledger integration fetches the token-list and signing
information. When the user wishes to transfer tokens, the wallet provides
Ledger with the signed token and program metadata.

> The Solana App pushes a signature verification on the metadata using Ledger's
> native transaction model. If the signature verification passes, then the program
> id and token mint information are valid, and the actual transfer proceeds. If
> it fails due to spoofed program or mint data, the whole transaction aborts.

Reviewer 3: ```suggestion
Ledger host applications provide token-list entry metadata is pushed to the device
via a new Ledger APDU command accepting **TBD_DATA_FORMAT** signed by the
token-list authority, whose public key is embedded in the Ledger app binary. Entries
whose signature verification passes are stored in a registry on the device that is
then referenced by a subsequent SIGN_SOLANA_TRANSACTION APDU command.
```

Author: Ah I see, I didn't get the part about an on-device registry.  Thanks for the clarification and the additional Ledger-related info :smile:

> TODO any associated token program considerations?

Reviewer 3: This should be compatible with support for the current ATA implementation, so long as the instruction format doesn't change

all over the code?

This solution creates friction for almost all SPL token users: on-chain programs,
wallets, applications, and end-users.

## Proposed Solution

Following the approach taken by Aave, we can lean on a crucial feature of
interest-bearing tokens: although the holding's token amount is constantly
changing, each token holder's proportion of total supply stays the same.

Therefore, we can satisfy the fungibility requirement for tokens on proportional
ownership. The token interface stays exactly the same, operating on shares instead
of tokens, and the concept of "token amount" is actually the "UI amount" to be
displayed by frontends.

Let's go through an example. Imagine we have an interest-bearing token mint with a total of
10,000 proportional shares, a 20% annual interest rate, and 1 share is currently
worth 100 tokens. If I transfer 1,000 tokens today, the program converts that
using the interest-bearing token program and asks it to move
10 shares. One year later, 1 share is worth 120 tokens, so if I transfer
1,000 tokens, the program converts and asks to move 8.33 shares. The same concept
applies to mints and burns.

> Following this model, we never have to worry about compounding interest. We

Reviewer 8: Would the interest be compounded periodically by the runtime and stored as a number in the i-token? Or would it need to be computed every time by `AmountToUiAmount` and not stored?

Author: The runtime would not be involved at all for this, so it would need to be computed every time and not stored anywhere.

useful for creating a holding when you don't know the program that you're
interacting with. See the Associated Token Account section for how to use this.

There are also new read-only instructions, which write data to a provided account
buffer:

* `AmountToUiAmount`: convert the given share amount to a UI token amount
* `UiAmountToAmount`: convert the given UI token amount to an exact share amount

See the Runtime section for more information on how these are used.

Programs and wallets that wish to support i-tokens must update the instructions
to use all `Checked` variants and use the new read-only instructions to fetch
information.

#### Conversions and Wrapper Library

The SPL token library provides wrappers compatible with any conforming program.
The wrapper simply checks the program id and deserializes the account data
as an `spl_token::Holding` or `spl_token::Mint`. If a client program requires access
to data specific to an i-token program, it needs to use that program's serializers.

To convert between amount and UI amount, the wrapper library calls the `AmountToUiAmount`
instruction on the i-token program, or for SPL token accounts, it calls

> `spl_token::amount_to_ui_amount` inline.

Reviewer: just fyi, I leaned that wstETH (wrapper version of stETH) just exists for the distinction of value-accruing vs re-basing token differences to be used for separately depending on upper defi primitive's token accounting assumption: https://help.lido.fi/en/articles/5231836-what-is-wrapped-steth-wsteth

Also, I think we now have return data runtime mechanism. :)

For interest-bearing tokens, after deserialization into a `spl_token::Holding`,
the `amount` field will be in terms of shares, and not token amount, so UIs and

> on-chain programs that need token amounts must convert using `AmountToUiAmount`.

Reviewer: stale paragraph?

Author: Michael and I had a long discussion about this part, so I hope it isn't stale :laughing:

We ended up deciding that `amount` is in terms of shares, so dapps / programs need to be aware that `ui_amount` can potentially be totally different due to a calculation during the `AmountToUiAmount` conversion, and know explicitly when to use `amount` vs `ui_amount`.  This makes interest-bearing tokens look just like normal tokens.  We still need some concrete feedback from the issuers of interest-bearing tokens to make sure this works for their system, but it also works much more easily for bridges.

Reviewer: I see! thanks for the clarification.

as a normal SPL token holding. The following byte must be the type of
account (ie. `Holding`), and after that, any data is allowed as required by the
new token program.

The same applies for mints, but only for the first 82 bytes.

### Token Program Registry

We need the [token-list](https://github.com/solana-labs/token-list) to include
vetted SPL token-conforming programs and group known mints by their program id.

The token list will publish a file of known program ids at
[token-list releases](https://github.com/solana-labs/token-list/releases/latest),
to be used by RPC and programs.

The entries in the token-list are signed using a managed key. As wallets or
Ledgers are asked to send transactions using a particular token program, they can
cross-check the mint or program information using a signature check on the 
token-list data.

The token-list client libraries can fetch, cache, and query all registry info,
to be used for simple program and mint validation.

TODO consider detailing the process of including a new program

TODO figure out how the token-list signing key will be managed

### Runtime

#### Ephemeral Accounts

> In order to reduce friction when using read-only instructions on other programs,

Reviewer 7: I think that having "set return value" and "get return value" syscalls would be a better approach and a better UX for devs designing programs which return computed information. Tag @jackcmay

Reviewer: oh, yeah. I had same idea for awhile. :) cc: @joncinque

Reviewer 9: I really like this idea and it would be very useful outside of interest bearing tokens. For example, it would allow CPI in Anchor to look like a normal function call.

Reviewer 2: @seanyoung, @jackcmay, dmakarov  and I had a chat about this earlier today.  @seanyoung is going to write up a brief design proposal PR soon.  Basically we need the same for solang as well to return arbitrary data to the CPI caller

we need the ability for on-chain programs to dynamically create "ephemeral
accounts", which only contain data, can be passed through CPI to other programs,
and disappear after the instruction.

For a program using i-tokens, to get the UI amount of a holding, the program creates
an ephemeral account with 8 bytes of data, passes the `amount` to the i-token program's
`AmountToUiAmount` instruction, then deserializes the data back into an `f64`.
At the end of the instruction, the account disappears.

This solution avoids the need to pass additional scratch accounts to almost all
programs that use tokens.

TODO what limits do we consider for these? A certain number of accounts?
A total amount of bytes? Regarding security, this breaks the previous
check that a CPI must use a subset of accounts provided to the calling program.

#### Dynamic sysvars

The interest-bearing token program also requires the use of dynamic sysvars. For example,
during the `AmountToUiAmount` instruction, the program needs to know the current

> time in order to properly convert from token amount to share amount.

Author: Thanks for noticing!

### Ledger

Ledger integrates with token-list in order to provide verified token program ids
and token tickers to the Ledger SDK and eliminate the need to store static
information in the Ledger Solana App.

A wallet or CLI with Ledger integration fetches the token-list and signing
information. When the user wishes to transfer tokens, the wallet first builds an
on-device token registry by pushing signed entries from the token-list for the subset
of tokens referenced by the transaction.

Ledger host applications provide token-list entry metadata is pushed to the device
via a new Ledger APDU command accepting **TBD_DATA_FORMAT** signed by the
token-list authority, whose public key is embedded in the Ledger app binary. Entries
whose signature verification passes are stored in a registry on the device that is
then referenced by a subsequent SIGN_SOLANA_TRANSACTION APDU command.

## Other New Token Programs

For awhile now, people have been coming up with ideas that likely require a new
token program to implement. These include:

### Voting Tokens

Outlined in a [GitHub issue](https://github.com/solana-labs/solana-program-library/issues/131),

> voting tokens use additional data about how long a token has been held.

Reviewer 10: The opposite logic would be to make it possible to lock tokens for certain amount of time to know for how long they would be held. This would allow to calculate voting power based on the long term commitment (skin in the game).
Users don't really like to deposit governance tokens to vote. It makes them very uncomfortable especially when the governance tokens they posses have substantial economic value. 
A better solution would be to allow the tokens to be locked to gain voting power while still being in user's possession.
